### PR TITLE
feat (api, worker): upload workflow artifact through temp url

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/workflow_queue_artifact.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
+package api
+
+import (
+	"context"
+	"mime"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/mux"
```
Comment (bnjjj) [inline engine/api/workflow_queue_artifact.go]: space

File: sdk/cdsclient/client_queue.go [hunk 2]
```diff
@@ -170,30 +172,133 @@ func (c *client) QueueSendResult(id int64, res sdk.Result) error {
...
+	retry = 50
+	for i := 0; i < retry; i++ {
+		uri := fmt.Sprintf("/queue/workflows/%d/artifact/%s/url/callback", id, tag)
+		_, callbackErr = c.PostJSON(uri, &art, nil)
+		if callbackErr == nil {
+			return nil
+		}
+	}
+
+	return globalErr
```
Comment (bnjjj) [inline sdk/cdsclient/client_queue.go]: s/globalErr/callbackErr

File: engine/api/workflow_queue_artifact.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+			WorkflowNodeRunID: nodeRun.ID,
+			WorkflowID:        nodeRun.WorkflowRunID,
+			Created:           time.Now(),
+		}
+
+		files := m.File[fileName]
+		if len(files) == 1 {
+			file, err := files[0].Open()
+			if err != nil {
+				return sdk.WrapError(err, "postWorkflowJobArtifactHandler> cannot open file")
```
Comment (sguiheux) [inline engine/api/workflow_queue_artifact.go]: file.close()

File: engine/api/workflow_queue_artifact.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+			}
+
+			if err := artifact.SaveWorkflowFile(&art, file); err != nil {
+				return sdk.WrapError(err, "postWorkflowJobArtifactHandler> Cannot save artifact in store")
```
Comment (sguiheux) [inline engine/api/workflow_queue_artifact.go]: file.close()

File: engine/api/workflow_queue_artifact.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		tag := vars["tag"]
+
+		hash, errG := generateHash()
+		if errG != nil {
+			return sdk.WrapError(errG, "postWorkflowJobArtifacWithTempURLtHandler> Could not generate hash")
+		}
+
+		art := sdk.WorkflowNodeRunArtifact{}
+		if err := UnmarshalBody(r, &art); err != nil {
+			return err
```
Comment (sguiheux) [inline engine/api/workflow_queue_artifact.go]: wrap

File: engine/api/workflow_queue_artifact.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+		store, ok := objectstore.Storage().(objectstore.DriverWithRedirect)
+		if !ok {
+			return sdk.WrapError(sdk.ErrForbidden, "postWorkflowJobArtifacWithTempURLtHandler > cast error")
+		}
+
+		// Load  Existing workflow Run Job
+		id, errI := requestVarInt(r, "permID")
+		if errI != nil {
+			return sdk.WrapError(sdk.ErrInvalidID, "postWorkflowJobArtifacWithTempURLtHandler> Invalid node job run ID")
```
Comment (sguiheux) [inline engine/api/workflow_queue_artifact.go]: errI

**Pull request merged**